### PR TITLE
Make json collapsible

### PR DESCRIPTION
**Pull request by smithjessk**
Make arrays and objects collapsible in the input and output fields. Implemented by updating `DetailRenderer`.

## Sneak Previews

image

image

## Details

- Thanks for the help @alexmojaki :) 
- A lot of the diff comes from migrated code.
- I didn't make super-meaningful commit messages. (Happy to change this pattern in the future; it's just not what i do most habitually based on prior work experience. I think reviewing at the level of PRs is easier.)
- I'm pretty unsure how idiomatic this code is. Again, happy for pushback.
- Also added `ui/storybook-static/` to `.gitignore`. Not sure what its point is- this might be a mistake! :) 
- I moved the call-location of `useToast`. I don't _think_ that  affects anything, assuming it's referentially transparent, but sorry if it isn't :/

### CODE REVIEW

File: ui/components/TracePage/DetailRenderer.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+interface ArrayRendererState {
+  expanded: boolean[];
+}
+
+class ArrayRenderer extends Component<ArrayRendererProps, ArrayRendererState> {
+  constructor(props: ArrayRendererProps) {
+    super(props);
+    this.state = {
```
Comment (alexmojaki) [inline ui/components/TracePage/DetailRenderer.tsx:41]: Please switch to a function-based component and use the `setState` hook to manage `expanded`.

Comment (alexmojaki) [reply ui/components/TracePage/DetailRenderer.tsx:41]: Same for the other class-based components.

**Comment by alexmojaki** [comment]
The storybook looks fine, but actually viewing a trace often goes wrong. Run `npm run build`, make a bunch of traces with `pytest` (maybe `rm -rf ~/.ought-ice/traces` first), and then try viewing a bunch of traces. In many cases I get an error which links to https://reactjs.org/docs/error-decoder.html?invariant=310 saying:

> Rendered more hooks than during the previous render.

The cause of the problem is probably this:

> I moved the call-location of useToast. I don't think that affects anything, assuming it's referentially transparent, but sorry if it isn't :/

Indeed, that's not allowed:

> Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns.

https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

If you find yourself repeatedly building the UI to debug something that can't be reproduced with storybook, I think this will have to wait until the dev server is fixed (cc @lslunis)

**Comment by smithjessk** [comment]
I just internalized that collapsing something resets the collapsing status of its children. Which, I think, is likely fine in the vast majority of cases, but I just wanted to draw attention to it!

**Comment by smithjessk** [comment]
Would it be better to change the button & the icon to be different depending on if the field is collapsed or not?

**Comment by smithjessk** [comment]
> The storybook looks fine, but actually viewing a trace often goes wrong. Run `npm run build`, make a bunch of traces with `pytest` (maybe `rm -rf ~/.ought-ice/traces` first), and then try viewing a bunch of traces. In many cases I get an error which links to https://reactjs.org/docs/error-decoder.html?invariant=310 saying:
> 
> > Rendered more hooks than during the previous render.
> 
> The cause of the problem is probably this:
> 
> > I moved the call-location of useToast. I don't think that affects anything, assuming it's referentially transparent, but sorry if it isn't :/
> 
> Indeed, that's not allowed:
> 
> > Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns.
> 
> https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level
> 
> If you find yourself repeatedly building the UI to debug something that can't be reproduced with storybook, I think this will have to wait until the dev server is fixed (cc @lslunis)

@alexmojaki Thanks! I think a70a48a0479663e06303675ded799b64e07b5850 addresses it.

**Comment by smithjessk** [comment]
> The storybook looks fine, but actually viewing a trace often goes wrong. Run `npm run build`, make a bunch of traces with `pytest` (maybe `rm -rf ~/.ought-ice/traces` first), and then try viewing a bunch of traces. In many cases I get an error which links to https://reactjs.org/docs/error-decoder.html?invariant=310 saying:
> 
> > Rendered more hooks than during the previous render.
> 
> The cause of the problem is probably this:
> 
> > I moved the call-location of useToast. I don't think that affects anything, assuming it's referentially transparent, but sorry if it isn't :/
> 
> Indeed, that's not allowed:
> 
> > Don’t call Hooks inside loops, conditions, or nested functions. Instead, always use Hooks at the top level of your React function, before any early returns.
> 
> https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level
> 
> If you find yourself repeatedly building the UI to debug something that can't be reproduced with storybook, I think this will have to wait until the dev server is fixed (cc @lslunis)

@alexmojaki Thanks! I think a70a48a0479663e06303675ded799b64e07b5850 addresses it.

File: ui/components/TracePage/DetailRenderer.tsx (hunk 1 above)
Comment (smithjessk) [reply ui/components/TracePage/DetailRenderer.tsx]: Thanks! https://github.com/oughtinc/ice/commit/a70a48a0479663e06303675ded799b64e07b5850 and f0a4b75703585942aafda7efdc1efd27641d324d address this.

**Comment by alexmojaki** [comment]
> Would it be better to change the button & the icon to be different depending on if the field is collapsed or not?

Definitely.

Something is still very wrong:

https://user-images.githubusercontent.com/3627481/214820706-b25bf23b-8afa-4bba-98c8-e21bd730697f.mp4

I think we need to wait until the dev server is fixed by @lslunis. Or you could try the partial fix in https://github.com/oughtinc/ice/pull/223 for the sake of development.

**Comment by smithjessk** [comment]
Mmmm thanks for the video capture. I can reproduce this locally so I will try to fix it tn!

**Comment by smithjessk** [comment]
@alexmojaki I found _a_ solution. Heavy caveat that it's just a proof of concept to demonstrate the shape of a solution to the bug.

The story:

- One bug is in the computation of the state `isExpanded` inside `ObjectRenderer`. 
- When I explicitly re-compute that state when the keys are [different lengths](https://github.com/oughtinc/ice/pull/226/commits/4211a1b5d3be32c736d1124f3e2b16d325c78f91#diff-0c65884279b027eab9e9bf87d3cd002f854d8656ffa0e331b749ddc7b4c1a3b1R103) [1], the bug goes away! 

https://user-images.githubusercontent.com/5957838/215015625-5a9d28f8-1755-43fb-aa81-dbf4cccdb739.mov

- The reason that no _values_ displayed, but the _keys_ did, is that the values being displayed depended on `isExpanded[key]`, which often returned false.

Things this story doesn't address (yet):
- why doesn't this failure seem to affect the rendering of the outputs?
- why can't I reproduce this bug with a script like this?

```python
from ice.recipe import recipe

async def f(x):
    return { "a": x }

async def answer():
    return await f({ "b": 1 })

recipe.main(answer)
```

- why are the keys seemingly coming from a parent value?
image


[1]: I stumbled across the 'different lengths' condition when debugging. Clearly this isn't actually the right recomputation condition 😅

**Comment by smithjessk** [comment]
As for next steps: I'm sure there's either 1) some other bug causing this, or 2) a more idiomatic way to handle cases like this. But I'm not sure about point 2 which affects my estimate for the likelihood of 1 :)

**Comment by stuhlmueller** [comment]
Closing for now, feel free to reopen.

**Pull request closed without merging**